src/aspire/denoising/denoiser_cov2d.py

```diff
@@ -18,72 +15,73 @@ class DenoiserCov2D(Denoiser):
         """
         super().__init__(src)
-        self.imgs_noise = self.src.images(start=0, num=np.inf).asnumpy()
         self.var_noise = var_noise
         if not isinstance(basis, FFBBasis2D):
             raise NotImplementedError('Currently only fast FB method is supported')
-        self.coeffs_noise = None
-        self.coeffs_estim = None
-        self.ctf_idx = None
-        self.ctf_fb = None
         self.basis = basis
-        self.imgs_estim = None
-
-        # Assign the CTF information and index for each image
-        if ctf_info and not (src.filters is None):
-            logger.info(f'Represent CTF filters in FB basis')
-            uniq_ctf_filters = list(set(src.filters))
-            # Create the indices of CTF filters from all images
-            self.ctf_idx = np.array([uniq_ctf_filters.index(f) for f in src.filters])
-            # Evaluate CTFs in the FFB basis
-            self.ctf_fb = [f.fb_mat(basis) for f in uniq_ctf_filters]
-        else:
-            logger.info(f'CTF filters are not included in Cov2D denoising')
-            # set all CTF filters to an identity filter
-            self.ctf_idx = np.zeros(self.nimg, dtype=int)
-            self.ctf_fb = [blk_diag_eye(blk_diag_partition(RadialCTFFilter().fb_mat(basis)))]
-
-    def denoise(self, covar_opt=None):
+        self.cov2d = None
+        self.mean_est = None
+        self.covar_est = None
+
+    def denoise(self, covar_opt=None, batch_size=512):
         """
-        Denoiser 2D images using Cov2D method
+         Build covariance matrix of 2D images and return a new ImageSource object
 
-        :param basis: The basis method to expand 2D images
         :param covar_opt: The option list for building Cov2D matrix
+        :param batch_size: The batch size for processing images
+        :return: A `DenoisedImageSource` object with the specified denoising object
         """
 
-        self.coeffs_noise = self.basis.evaluate_t(self.imgs_noise)
-
-        cov2d = RotCov2D(self.basis)
+        # Initialize the rotationally invariant covariance matrix of 2D images
+        # A fixed batch size is used to go through each image
+        self.cov2d = BatchedRotCov2D(self.src, self.basis, batch_size=batch_size)
 
         default_opt = {'shrinker': 'frobenius_norm', 'verbose': 0, 'max_iter': 250,
             'iter_callback': [], 'store_iterates': False, 'rel_tolerance': 1e-12,
             'precision': 'float64'}
         covar_opt = fill_struct(covar_opt, default_opt)
-        mean_coeffs_est = cov2d.get_mean(self.coeffs_noise, self.ctf_fb, self.ctf_idx)
-
-        covar_coeffs_est = cov2d.get_covar(self.coeffs_noise, self.ctf_fb, self.ctf_idx, mean_coeffs_est,
-                                               noise_var=self.var_noise, covar_est_opt=covar_opt)
+        # Calculate the mean and covariance for the rotationally invariant covariance matrix of 2D images
+        self.mean__est = self.cov2d.get_mean()
```

Comment:
Shouldn't this be `mean_est`? Does this not cause the code to crash?

Reply:
Good catch. It will recalculate it if the value is None.